soc/nordic/common/dmm.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * nRF SoC specific public APIs for Device Memory Management (dmm) subsystem
+ */
+
+#ifndef SOC_NORDIC_COMMON_DMM_H_
+#define SOC_NORDIC_COMMON_DMM_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @cond INTERNAL_HIDDEN */
+
+/**
+ * @brief Prepare a DMA output buffer for the specified device
+ *
+ * Allocate an output buffer in memory region that given device can perform DMA transfers from.
+ * Copy @p user_buffer contents into it.
+ * Writeback data cache lines associated with output buffer, if needed.
+ *
+ * @note Depending on provided user buffer parameters and SoC architecture,
+ *       dynamic allocation and cache operations might be skipped.
+ *
+ * @warning It is prohibited to read or write @p user_buffer or @p buffer_out contents
+ *          until @ref dmm_dma_buffer_out_release() is called
+ *
+ * @param dev Device to prepare the buffer for.
+ * @param user_buffer Buffer containing data to be processed by the given device
+ * @param user_out Pointer to a pointer to be filled by prepared DMA buffer address.
+ *
+ * @retval 0 If succeeded.
+ * @retval -ENOMEM If output buffer could not be allocated.
+ * @retval -errno Negative errno for other failures.
+ */
+int dmm_dma_buffer_out_prepare(const struct device *dev, void const * user_buffer, void ** buffer_out);
+
+/**
+ * @brief Release the previously prepared DMA output buffer
+ *
+ * @param dev Device to release the buffer for.
+ * @param buffer_out DMA output buffer previously prepared with @ref dmm_dma_buffer_out_prepare().
+ *
+ * @retval 0 If succeeded.
+ * @retval -errno Negative errno code on failure.
+ */
+int dmm_dma_buffer_out_release(const struct device *dev, void const * buffer_out);
+
+/**
+ * @brief Prepare a DMA input buffer for the specified device
+ *
+ * Allocate an input buffer in memory region that given device can perform DMA transfers to.
+ *
+ * @note Depending on provided user buffer parameters and SoC architecture,
+ *       dynamic allocation might be skipped.
+ *
+ * @warning It is prohibited to read or write @p user_buffer or @p buffer_in contents
+ *          until @ref dmm_dma_buffer_in_release() is called
+ *
+ * @param dev Device to prepare the buffer for.
+ * @param user_buffer Buffer to be filled with data from the given device
+ * @param user_in Pointer to a pointer to be filled by prepared DMA buffer address.
+ *
+ * @retval 0 If succeeded.
+ * @retval -ENOMEM If input buffer could not be allocated.
+ * @retval -errno Negative errno for other failures.
+ */
+int dmm_dma_buffer_in_prepare(const struct device *dev, void const * user_buffer, void ** buffer_in);
```

Comment:
since in `dmm_dma_buffer_in_release` you are also providing `user_buffer` then it seems that `user_buffer` is redundant here.

Reply:
the thing is that `dmm_dma_buffer_in_prepare` needs to check whether the provided buffer has correct attributes (location, alignment) before performing dynamic allocation:
- if `user_buffer` is in correct location, reachable by DMA of given device, and has correct attributes, then `*buffer_in = user_buffer;`
- otherwise `*buffer_in = dynamically_allocate(user_length);`